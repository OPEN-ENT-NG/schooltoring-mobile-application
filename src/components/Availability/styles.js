import { StyleSheet } from "react-native";

export default StyleSheet.create({
  container: {
    paddingHorizontal: 30,
    paddingVertical: 15,
    flex: 1,
    justifyContent: "space-evenly"
  },
  item: {
    flexDirection: "row",
    justifyContent: "flex-start",
    alignItems: "center",
    height: 60
  },
  itemSwitch: {
    marginRight: 20
  },
  itemText: {
    fontSize: 18
  },
  list: {
    flex: 1,
    justifyContent: "space-around"
  },
  title: {
    fontSize: 18,
    marginBottom: 40,
    flex: 0
  },
  buttonView: {
    flexDirection: "row",
    justifyContent: "flex-end",
    flex: 0,
    padding: 15
  }
});
