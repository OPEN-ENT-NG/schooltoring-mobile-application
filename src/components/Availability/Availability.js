import React, { Component } from "react";
import {
  View,
  Text,
  Switch,
  TouchableWithoutFeedback,
  ScrollView
} from "react-native";
import PropTypes from "prop-types";

import styles from "./styles";

import I18n from "../../api/I18n";

import SecondaryButton from "../SecondaryButton/SecondaryButton";
import Header from "../Header/Header";

export default class Availability extends Component {
  static navigationOptions = ({ navigation }) => ({
    header: (
      <Header
        navigation={navigation}
        title={I18n.t(`${navigation.state.routeName.toLowerCase()}.title`)}
      />
    )
  });

  constructor(props) {
    super(props);
    this.state = props.screenProps.profile.availabilities || {
      monday: true,
      tuesday: true,
      wednesday: true,
      thursday: true,
      friday: true,
      saturday: true,
      sunday: true
    };

    this.renderDay = this.renderDay.bind(this);
    this.setDayValue = this.setDayValue.bind(this);
  }

  setDayValue(day, value) {
    let state = {};
    state[day] = value;
    this.setState(state);
  }

  renderDay(day) {
    return (
      <View key={day} style={styles.item}>
        <Switch
          style={styles.itemSwitch}
          value={this.state[day]}
          onValueChange={value => {
            this.setDayValue(day, value);
          }}
        />
        <TouchableWithoutFeedback
          hitSlop={{
            top: 20,
            left: 20,
            right: 40,
            bottom: 20
          }}
          onPress={() => {
            this.setDayValue(day, !this.state[day]);
          }}
        >
          <View>
            <Text style={styles.itemText}>{I18n.t(`availability.${day}`)}</Text>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <View style={styles.container}>
          <Text style={styles.title}>{I18n.t("availability.description")}</Text>
          <ScrollView contentContainerStyle={styles.list}>
            {Object.keys(this.state).map(day => this.renderDay(day))}
          </ScrollView>
        </View>
        <View style={styles.buttonView}>
          <SecondaryButton
            style={styles.button}
            onPress={() => {
              this.props.screenProps.onChangeScreen("availabilities", {
                monday: false,
                tuesday: false,
                wednesday: false,
                thursday: false,
                friday: false,
                saturday: false,
                sunday: false
              });
              this.props.navigation.push("Profile");
            }}
            title={I18n.t("skip")}
          />
          <SecondaryButton
            style={styles.button}
            onPress={() => {
              this.props.screenProps.onChangeScreen(
                "availabilities",
                this.state
              );
              this.props.navigation.push("Profile");
            }}
            title={I18n.t("next")}
          />
        </View>
      </View>
    );
  }
}
