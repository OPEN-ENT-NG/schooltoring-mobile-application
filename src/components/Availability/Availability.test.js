import React from "react";
import { Switch } from "react-native";
import { shallow } from "enzyme";

import Availability from "./Availability";

import state from "../../jest/state.json";

let newState;
const navigation = {
  push: jest.fn(),
  pop: jest.fn()
};
let onChangeScreen = jest.fn().mockImplementation((screen, state) => {
  newState = state;
});

const rendered = shallow(
    <Availability
      {...state}
      navigation={navigation}
      onChangeScreen={onChangeScreen}
    />
  ),
  switches = rendered.find(Switch),
  days = Object.keys(state.profile.availabilities),
  renderedWithSaveButton = shallow(
    <Availability
      {...state}
      navigation={navigation}
      onChangeScreen={onChangeScreen}
      saveButton={true}
    />
  );

describe("Render", () => {
  test("Availability without save button should render without crashing", () => {
    expect(rendered).toMatchSnapshot();
  });

  test("Availability with save button should render without crashing", () => {
    expect(renderedWithSaveButton).toMatchSnapshot();
  });

  test("Availaibility without save button should render 2 Secondary Buttons", () => {
    expect(rendered.find("SecondaryButton").length).toEqual(2);
  });

  test("Availability with save button should render only 1 Secondary Button", () => {
    expect(renderedWithSaveButton.find("SecondaryButton").length).toEqual(1);
  });
  test(`${days.length} switch should be rendered`, () => {
    expect(switches.length).toEqual(days.length);
  });

  test("Switches values should match with state values", () => {
    for (let i = 0; i < days.length; i++) {
      expect(switches.at(i).props().value).toEqual(
        state.profile.availabilities[days[i]]
      );
    }
  });
});

describe("Interactions", () => {
  const skipButton = rendered.find("SecondaryButton").at(0);
  const nextButton = rendered.find("SecondaryButton").at(1);
  const saveButton = renderedWithSaveButton.find("SecondaryButton").first();

  test("Changing switch values should update state", () => {
    let item, newState;
    for (let i = 0; i < switches.length; i++) {
      item = switches.at(i);
      item.props().value = !item.props().value;
      item.props().onValueChange();
    }
    newState = rendered.update().state();
    for (let i = 0; i < days.length; i++) {
      expect(newState[days[i]]).not.toEqual(
        state.profile.availabilities[days[i]]
      );
    }
  });

  test("Press 'Next' button should trigger onChangeScreen function and navigation.push function", () => {
    nextButton.props().onPress();
    expect(onChangeScreen).toHaveBeenCalled();
    expect(navigation.push).toHaveBeenCalled();
  });

  test("Press 'Skip' button should trigger onChangeScreen function and navigation.push function", () => {
    onChangeScreen.mockReset();
    navigation.push.mockReset();
    skipButton.props().onPress();

    expect(onChangeScreen).toHaveBeenCalled();
    expect(navigation.push).toHaveBeenCalled();
  });

  test("Press 'Skip' button should set all days as false", () => {
    onChangeScreen.mockReset();
    skipButton.props().onPress();
    days.forEach(day => {
      expect(newState[day]).toBeFalsy();
    });
  });

  test("Press 'Save' button should trigger onChangeScreen function and navigator.pop function", () => {
    onChangeScreen.mockReset();
    navigation.pop.mockReset();
    saveButton.props().onPress();

    expect(onChangeScreen).toHaveBeenCalled();
    expect(navigation.pop).toHaveBeenCalled();
  });
});
