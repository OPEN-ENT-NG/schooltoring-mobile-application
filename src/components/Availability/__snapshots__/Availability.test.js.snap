// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Availability with save button should render without crashing 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-evenly",
        "paddingHorizontal": 30,
        "paddingVertical": 15,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#232323",
          "flex": 0,
          "fontSize": 18,
          "marginBottom": 40,
        }
      }
    >
      Les jours où je suis le plus disponible
    </Text>
    <ScrollView
      contentContainerStyle={
        Object {
          "flex": 1,
          "justifyContent": "space-around",
        }
      }
    >
      <Component
        key="monday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Lundi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="tuesday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={true}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Mardi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="wednesday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Mercredi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="thursday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={true}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Jeudi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="friday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Vendredi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="saturday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Samedi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="sunday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={true}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Dimanche
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
    </ScrollView>
  </Component>
  <Component
    style={
      Object {
        "flex": 0,
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "padding": 15,
      }
    }
  >
    <SecondaryButton
      onPress={[Function]}
      title="ENREGISTRER"
    />
  </Component>
</Component>
`;

exports[`Render Availability without save button should render without crashing 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-evenly",
        "paddingHorizontal": 30,
        "paddingVertical": 15,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#232323",
          "flex": 0,
          "fontSize": 18,
          "marginBottom": 40,
        }
      }
    >
      Les jours où je suis le plus disponible
    </Text>
    <ScrollView
      contentContainerStyle={
        Object {
          "flex": 1,
          "justifyContent": "space-around",
        }
      }
    >
      <Component
        key="monday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Lundi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="tuesday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={true}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Mardi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="wednesday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Mercredi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="thursday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={true}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Jeudi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="friday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Vendredi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="saturday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={false}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Samedi
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
      <Component
        key="sunday"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "flex-start",
          }
        }
      >
        <Switch
          disabled={false}
          onValueChange={[Function]}
          style={
            Object {
              "marginRight": 20,
            }
          }
          value={true}
        />
        <TouchableWithoutFeedback
          hitSlop={
            Object {
              "bottom": 20,
              "left": 20,
              "right": 40,
              "top": 20,
            }
          }
          onPress={[Function]}
        >
          <Component>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#232323",
                  "fontSize": 18,
                }
              }
            >
              Dimanche
            </Text>
          </Component>
        </TouchableWithoutFeedback>
      </Component>
    </ScrollView>
  </Component>
  <Component
    style={
      Object {
        "flex": 0,
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "padding": 15,
      }
    }
  >
    <SecondaryButton
      onPress={[Function]}
      title="PASSER"
    />
    <SecondaryButton
      onPress={[Function]}
      title="SUIVANT"
    />
  </Component>
</Component>
`;
